File: rust/phaserz/src/terrain.rs
use godot::global::{floori, randf};
use godot::prelude::*;
use godot::classes::{
    mesh::{PrimitiveType, ArrayType},
    fast_noise_lite::{FractalType, NoiseType},
    ArrayMesh,
    Node3D, INode3D,
    MeshInstance3D,
    Shader,
    ShaderMaterial,
    FastNoiseLite,
    ImageTexture
};

#[derive(GodotClass)]
#[class(base=Node3D)]
struct Terrain {
    #[export]
    divisions: i32,

    points: PackedVector3Array,
    indices: PackedInt32Array,
    max_point_width: i32,

    base: Base<Node3D>
}
#[godot_api]
impl INode3D for Terrain {
    fn init(base: Base<Node3D>) -> Self {
        Self {
            divisions: 8,
            points: PackedVector3Array::new(),
            indices: PackedInt32Array::new(),
            max_point_width: 0,
            base
        }
    }
    fn ready(&mut self) {
        let mut vertex_indices: Vec<Vec<i32>> = Vec::new();
        // Create vertices
        let min_x_num = (2 as i32).pow(self.divisions as u32) + 1;
        let max_x_num = (2 as i32).pow(self.divisions as u32 + 1) + 1;
        let z_num = max_x_num;
        let l = 128.0 / (max_x_num - 1) as f32;
        let h = 0.5 * l * f32::sqrt(3.0);
        let mut points_array: Vec<Vector3> = Vec::new();
        let mut index = 0;
        for z in 0..z_num {
            let x_num: i32;
            let second_half: bool;
            let second_half_z = z - (z_num / 2 + 1) + 1;
            if second_half_z >= 0 {
                x_num = max_x_num - second_half_z;
                second_half = true;
            } else {
                x_num = min_x_num + z;
                second_half = false;
            }
            let mut row_indices: Vec<i32> = Vec::new();
            for x in 0..x_num {
                let current_x: f32;
                if second_half {
                    current_x = -64.0 + 0.5 * l * second_half_z as f32 + l * x as f32;
                } else {
                    current_x = -32.0 - 0.5 * l * z as f32 + l * x as f32;
                }
                let current_z = 32.0 * f32::sqrt(3.0) - h * z as f32;
                points_array.push(Vector3::new(current_x, 0.0, current_z));
                row_indices.push(index);
                index += 1;
            }
            vertex_indices.push(row_indices);
        }
        self.points = points_array.into();
        self.max_point_width = max_x_num;
        // Create indices
        let mut indices: Vec<i32> = Vec::new();
        for z in 0..vertex_indices.len() - 1 {
            let top_row = &vertex_indices[z];
            let bottom_row = &vertex_indices[z + 1];
            let top_row_len = top_row.len();
            let bottom_row_len = bottom_row.len();
            if bottom_row_len > top_row_len {
                for x in 0..bottom_row_len + top_row_len - 2 {
                    let triangle: [i32; 3];
                    if x % 2 == 0 {
                        triangle = [bottom_row[x / 2], bottom_row[x / 2 + 1], top_row[x / 2]];
                    } else {
                        triangle = [top_row[x / 2], bottom_row[x / 2 + 1], top_row[x / 2 + 1]];
                    }
                    indices.extend(triangle);
                }
            } else {
                for x in 0..bottom_row_len + top_row_len - 2 {
                    let triangle: [i32; 3];
                    if x % 2 == 0 {
                        triangle = [top_row[x / 2], bottom_row[x / 2], top_row[x / 2 + 1]];
                    } else {
                        triangle = [bottom_row[x / 2], bottom_row[x / 2 + 1], top_row[x / 2 + 1]];
                    }
                    indices.extend(triangle);
                }
            }
        }
        self.indices = indices.into();
        // Create origin cell
        self.add_cell(Vector2i::ZERO);
    }
}
#[godot_api]
impl Terrain {
    fn add_cell(&mut self, position: Vector2i) {
        let mut arrays: Array<Variant> = Array::new();
        arrays.resize(ArrayType::MAX.ord() as usize, &Variant::nil());
        arrays.set(ArrayType::VERTEX.ord() as usize, &self.points.to_variant());
        arrays.set(ArrayType::INDEX.ord() as usize, &self.indices.to_variant());
        
        let mut material = ShaderMaterial::new_gd();
        material.set_shader(&(load("res://shaders/terrain.gdshader") as Gd<Shader>));
        let mut noise = FastNoiseLite::new_gd();
        noise.set_noise_type(NoiseType::SIMPLEX);
        noise.set_seed(floori(randf()) as i32);
        noise.set_frequency(0.01);
        noise.set_fractal_type(FractalType::FBM);
        noise.set_fractal_octaves(5);
        noise.set_fractal_lacunarity(2.0);
        noise.set_fractal_gain(0.5);
        let heightmap_img = noise.get_image(self.max_point_width, self.max_point_width).unwrap();
        let heightmap_tex = ImageTexture::create_from_image(&heightmap_img).unwrap();
        material.set_shader_parameter("heightmap", &heightmap_tex.to_variant());

        let mut mesh = ArrayMesh::new_gd();
        mesh.add_surface_from_arrays(PrimitiveType::TRIANGLES, &arrays);
        mesh.surface_set_material(0, &material);

        let mut mesh_instance = MeshInstance3D::new_alloc();
        mesh_instance.set_mesh(&mesh);

        // Give mesh instance shader material, add it to coordinate system, etc.
        self.base_mut().add_child(&mesh_instance);
    }
}

File: scripts/Terrain/Terrain.cs
using Godot;
using Godot.Collections;
using System;
using System.Collections.Generic;
using System.Linq;

public partial class Terrain : Node3D
{
	[Export]
	private int _divisions = 8;
	private Vector3[] _points;
	private int[] _indices;
	private int _maxPointWidth;

	// Called when the node enters the scene tree for the first time.
	public override void _Ready()
	{
		// INITIALIZING MESH GEOMETRY
		// Creating vertices
		List<List<int>> vertexIndices = [];
		int minXNum = (int) (Math.Pow(2, _divisions) + 1);
		int maxXNum = (int) (Math.Pow(2, _divisions + 1) + 1);
		var zNum = maxXNum;
		float l = 128f / (maxXNum - 1);
		float h = (float) (.5 * l * Math.Sqrt(3));
		List<Vector3> pointsArray = [];
		int index = 0;
		for(int z = 0; z < zNum; z++)
		{
			int xNum;
			bool secondHalf;
			int secondHalfZ = z - (zNum / 2 + 1) + 1;
			if(secondHalfZ >= 0)
			{
				xNum = maxXNum - secondHalfZ;
				secondHalf = true;
			}
			else
			{
				xNum = minXNum + z;
				secondHalf = false;
			}
			List<int> rowIndices = [];
			for(int x = 0; x < xNum; x++)
			{
				float currentX;
				if(secondHalf)
				{
					currentX = (float) (-64.0 + 0.5 * l * secondHalfZ + l * x);
				}
				else
				{
					currentX = (float) (-32.0 - 0.5 * l * z + l * x);
				}
				var currentZ = (float) (32 * Math.Sqrt(3) - h * z);
				pointsArray.Add(new Vector3(currentX, 0f, currentZ));
				rowIndices.Add(index++);
			}
			vertexIndices.Add(rowIndices);
		}
		_points = [.. pointsArray];
		_maxPointWidth = maxXNum;
		// Create indices
		List<int> indices = [];
		for(int z = 0; z < vertexIndices.Count - 1; z++)
		{
			var topRow = vertexIndices[z];
			var bottomRow = vertexIndices[z + 1];
			var topRowLen = topRow.Count;
			var bottomRowLen = bottomRow.Count;
			if(bottomRowLen > topRowLen)
			{
				for(int x = 0; x < bottomRowLen + topRowLen - 2; x++)
				{
					int[] triangle;
					if(x % 2 == 0)
					{
						triangle = [bottomRow[x / 2], bottomRow[x / 2 + 1], topRow[x / 2]];
					}
					else
					{
						triangle = [topRow[x / 2], bottomRow[x / 2 + 1], topRow[x / 2 + 1]];
					}
					indices.AddRange(triangle);
				}
			}
			else
			{
				for(int x = 0; x < bottomRowLen + topRowLen - 2; x++)
				{
					int[] triangle;
					if(x % 2 == 0)
					{
						triangle = [topRow[x / 2], bottomRow[x / 2], topRow[x / 2 + 1]];
					}
					else
					{
						triangle = [bottomRow[x / 2], bottomRow[x / 2 + 1], topRow[x / 2 + 1]];
					}
					indices.AddRange(triangle);
				}
			}
		}
		_indices = [.. indices];
		// Add origin cell
		AddCell(Vector2I.Zero);
	}
	public void AddCell(Vector2I position)
	{
		Array<Variant> arrays = [];
		arrays.Resize((int) Mesh.ArrayType.Max);
		arrays[(int) Mesh.ArrayType.Vertex] = _points;
		arrays[(int) Mesh.ArrayType.Index] = _indices;

        var material = new ShaderMaterial
        {
            Shader = GD.Load<Shader>("res://shaders/terrain.gdshader")
        };
        var noise = new FastNoiseLite
        {
            NoiseType = FastNoiseLite.NoiseTypeEnum.Perlin,
            Seed = (int) Math.Floor(GD.Randf()),
			Frequency = 0.01f,
			FractalType = FastNoiseLite.FractalTypeEnum.Fbm,
			FractalOctaves = 5,
			FractalLacunarity = 2.0f,
			FractalGain = 0.5f
        };
		var heightmapImg = noise.GetImage(_maxPointWidth, _maxPointWidth);

		// Blur heightmap
		// Create rendering device and shader
		var rd = RenderingServer.CreateLocalRenderingDevice();
        var shaderFile = GD.Load<RDShaderFile>("res://shaders/glsl/blur.glsl");
        var shaderSpirv = shaderFile.GetSpirV();
        var shader = rd.ShaderCreateFromSpirV(shaderSpirv);
		// Create uniforms
		var heightmapInputData = heightmapImg.GetData().Select(x => (int)x).ToArray();
		var heightmapBytes = new byte[heightmapInputData.Length * sizeof(int)];
		Buffer.BlockCopy(heightmapInputData, 0, heightmapBytes, 0, heightmapBytes.Length);
		var inputBuffer = rd.StorageBufferCreate((uint) heightmapBytes.Length, heightmapBytes);
		var inputUniform = new RDUniform()
		{
			UniformType = RenderingDevice.UniformType.StorageBuffer,
			Binding = 0
		};
		inputUniform.AddId(inputBuffer);
		var outputData = new byte[heightmapInputData.Length * sizeof(int)];
		System.Array.Fill(outputData, (byte) 0);
		var outputBuffer = rd.StorageBufferCreate((uint) heightmapBytes.Length, outputData);
		var outputUniform = new RDUniform()
		{
			UniformType = RenderingDevice.UniformType.StorageBuffer,
			Binding = 1
		};
		outputUniform.AddId(outputBuffer);
		int[] sizeInt = [_maxPointWidth];
		var sizeData = new byte[sizeof(int)];
		Buffer.BlockCopy(sizeInt, 0, sizeData, 0, sizeData.Length);
		var sizeBuffer = rd.StorageBufferCreate(sizeof(int), sizeData);
		var sizeUniform = new RDUniform()
		{
			UniformType = RenderingDevice.UniformType.StorageBuffer,
			Binding = 2
		};
		sizeUniform.AddId(sizeBuffer);
		int[] radiusInt = [4];
		var radiusData = new byte[sizeof(int)];
		Buffer.BlockCopy(radiusInt, 0, radiusData, 0, radiusData.Length);
		var radiusBuffer = rd.StorageBufferCreate(sizeof(int), radiusData);
		var radiusUniform = new RDUniform()
		{
			UniformType = RenderingDevice.UniformType.StorageBuffer,
			Binding = 3
		};
		radiusUniform.AddId(radiusBuffer);
		var uniformSet = rd.UniformSetCreate(
			[
				inputUniform,
				outputUniform,
				sizeUniform,
				radiusUniform
			],
			shader,
			0
		);
		// Create compute pipeline
		var pipeline = rd.ComputePipelineCreate(shader);
		var computeList = rd.ComputeListBegin();
		rd.ComputeListBindComputePipeline(computeList, pipeline);
		rd.ComputeListBindUniformSet(computeList, uniformSet, 0);
		// FIXME: this isn't size-dependent.
		rd.ComputeListDispatch(computeList, xGroups: 512/16, yGroups: 512/16, zGroups: 1);
		rd.ComputeListEnd();
		// Run blur
		rd.Submit();
		rd.Sync();

		// Finish
		// Create texture
		var outputBytes = rd.BufferGetData(outputBuffer);
		var blurredInts = new int[heightmapInputData.Length];
		Buffer.BlockCopy(outputBytes, 0, blurredInts, 0, heightmapInputData.Length);
		var blurredImageData = blurredInts.Select(x => (byte)x).ToArray();
		var blurredImage = Image.CreateFromData(_maxPointWidth, _maxPointWidth, false, Image.Format.L8, blurredImageData);
		var blurredTex = ImageTexture.CreateFromImage(blurredImage);
		material.SetShaderParameter("heightmap", blurredTex);
		// Create mesh
		var mesh = new ArrayMesh();
		mesh.AddSurfaceFromArrays(Mesh.PrimitiveType.Triangles, (Godot.Collections.Array) arrays);
		mesh.SurfaceSetMaterial(0, material);
        var meshInstance = new MeshInstance3D
        {
            Mesh = mesh
        };
		// Add child
		AddChild(meshInstance);
    }
}
